e2e: framework: clean up and export NUMA alignment utilites

ffromani opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
This is a planned and somehow needed cleanup for the e2e node topology manager tests.
There is room for further cleanup in this area (e.g. cpumanager and topology manager code deduplication) considering the work which is planned in this area.

We push to the framework code which was meant from the beginning to be reusable, and we clean up the test package. This is also an enabler for more tests.

**Which issue(s) this PR fixes**:
Fixes: N/A

**Special notes for your reviewer**:
Internal cleanup only: no intended change in external behaviour.

**Does this PR introduce a user-facing change?**:
```
NONE
```

fromanirh commented:
/sig node

fromanirh commented:
/sig testing

fromanirh commented:
/assign @timothysc

neolit123 commented:
this change and the addition of unit tests requires a reviewer from SIG Node.
try asking in #sig-node on slack once your PR is no longer WIP.

/unassign @timothysc

fromanirh commented:
basic verification passed, pending final verification. This PR is reviewable

oomichi commented:
/ok-to-test
/release-note-none
/cc @oomichi

fromanirh commented:
Thanks for the ok-to-test! I'll bring this up in the sig-node forum. Still waiting to get hw access again and do the final verification - the change is reviewable and works as expected on single-numa-node boxes. Now let me check/fix the CI failures.

fromanirh commented:
/retest

oomichi commented on test/e2e_node/numa_alignment.go:
```diff
@@ -1,273 +0,0 @@
-/*
```
Why do we need to move this package into e2e test framework(`test/e2e/framework/`)?
Those functions are called from test/e2e_node tests only, so it is nice to keep it here for related functions.

fromanirh replied:
The only reason is I thought `test/e2e/framework/node` was the correct place where to put reusable code. I'll move as you suggested

oomichi commented on test/e2e/framework/node/cpuset.go:
```diff
@@ -0,0 +1,69 @@
+/*
```
Why don't we implement this package under ` test/e2e_node/`?
Those functions seem to be used in ` test/e2e_node/` only.

fromanirh replied:
Thanks for raising this. I can move back under `test/e2e_node`, no problem.
In general is worth noting that this package actually duplicates kubelet code (https://github.com/kubernetes/kubernetes/tree/master/pkg/kubelet/cm/cpuset), this is because AFAIU we should not use kubelet production code in the test implementation itself, so a little duplication is better. Is my understanding right?

oomichi replied:
We have moved test specific functions from e2e test framework (test/e2e/framework/) already with much effort as https://github.com/kubernetes/kubernetes/issues/84380 I'd like to avoid such work again.
e2e test framework should not depend on the other packages in kubernetes/kubernetes repo.
Your understanding seems good for me, but if we move those functions from the framework, we don't need to make duplication because of the outside of the framework.

fromanirh replied:
Thanks for your feedback. I'll updated a much narrower PR soon. I still believe there is some merit reorganizing this code under a subpackage in `test/e2e_node`. But no more additions to the framework, and no more duplicated packages.

fromanirh commented:
uhm, the `verify` job reported failure yet in the details I see 28/28 tests passed. Let me just retrigger.

fromanirh commented:
/retest

fromanirh commented:
So I took the failing job `build log` and did ```bash
$ grep 'exit code' build-log.txt 
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
+++ exit code: 0
```
I guess I'll retry again.

fromanirh commented:
/retest

fromanirh commented:
/retest

oomichi commented on test/e2e_node/alignment/numa.go:
```diff
@@ -0,0 +1,177 @@
+	"strings"
+
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+
```
unnecessary blank here in the same `k8s.io/*` block.

fromanirh replied:
thanks, fixed.

oomichi commented on test/e2e_node/alignment/numa.go:
```diff
@@ -0,0 +1,177 @@
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
+
+	v1 "k8s.io/api/core/v1"
+
```
ditto

fromanirh replied:
thanks, fixed.

oomichi commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -35,26 +37,34 @@ import (
+	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
 	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
 	"k8s.io/kubernetes/pkg/kubelet/types"
+
```
ditto

fromanirh replied:
thanks, fixed.

oomichi commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -35,26 +37,34 @@ import (
 	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
 	e2eskipper "k8s.io/kubernetes/test/e2e/framework/skipper"
 	e2etestfiles "k8s.io/kubernetes/test/e2e/framework/testfiles"
 
```
ditto

fromanirh replied:
thanks, fixed.

oomichi commented:
kindly ping @fromanirh for come review comments.

fromanirh commented:
status update: first of all, thanks for the reviews! I still believe this patch should get in eventually, but can wait a bit more.
 I'd like to collect feedback from the sig-node CI/testing team currently being discussed.
There is discussion ongoing in the sig-node community about how reorganize the team after the inital push to improve the testing. I'll bring this patch up in the next meeting as soon as we agree on the schedule.

dashpole commented on test/e2e_node/alignment/numa.go:
```diff
@@ -0,0 +1,175 @@
+limitations under the License.
+*/
+
+package alignment
```
Just to confirm, did you use `git mv` to move this from test/e2e_node/numa_alignment.go.  It would be nice to see the diff and preserve history.

fromanirh replied:
Initially yes but then it got lost in rebase. Thanks for pointing this out, I'll fix and push again.

fromanirh replied:
The last upload explicitely address this point. `git log --follow` works as expected.

dashpole reviewed (commented):
/priority backlog

ffromani committed:
test: e2e: topomgr: move test-specific utilites

This patch prepares for the further generalization of
`numa_alignment.go` moving test-specific utilities, which
aren't yet useful to other tests, closer to their clients.

This way we enable a future patch to move `numa_alignment.go`
into a proper utility module.

No intended changes in behaviour.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: framework: clean/export NUMA alignment utils

We recently added topology manager e2e tests.
This pulled in some utilities to check the NUMA alignment
of the resources of a container, and these utilities can and
should be generalized and exported in the e2e framework.

Doing this will enable us to do further cleanup and consolidation
in this area, and make it easier to add more tests for both existing
components and for future extensions which are on the roadmap in the
future (example: memory manager).

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
e2e: node: topomgr: avoid plugin leak on test fail

We need to make sure we tear down the sriov device plugin pod
should the tests fail, to avoid leaking pods in the test environment.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

dashpole commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -21,34 +21,38 @@ import (
 
 const (
-	numalignCmd = `export CPULIST_ALLOWED=$( awk -F":\t*" '/Cpus_allowed_list/ { print $2 }' /proc/self/status); env; sleep 1d`
+	NUMAAlignmentSleepCommand = `export CPULIST_ALLOWED=$( awk -F":\t*" '/Cpus_allowed_list/ { print $2 }' /proc/self/status); env; sleep 1d`
```
It looks like this is the same as the constant in test/e2e_node/alignment/numa.go.  It would be nice to consolidate

dashpole commented on test/e2e_node/topology_manager_test.go:
```diff
@@ -297,19 +301,27 @@ func findSRIOVResource(node *v1.Node) (string, int64) {
-		if numaRes != nil {
-			framework.Logf("NUMA resources for %s/%s: %s", pod.Name, cnt.Name, numaRes.String())
+
+		cntFileRead := func(podName, cntName, path string) string {
```
Why define this function here and plumb it through?  Since this is the only call site, it seems like it just adds complexity.
